include/openthread/dataset.h

```diff
@@ -221,19 +221,19 @@ typedef struct otOperationalDatasetComponents
  */
 typedef struct otOperationalDataset
 {
-    uint64_t                       mActiveTimestamp;  ///< Active Timestamp
-    uint64_t                       mPendingTimestamp; ///< Pending Timestamp
-    otNetworkKey                   mNetworkKey;       ///< Network Key
-    otNetworkName                  mNetworkName;      ///< Network Name
-    otExtendedPanId                mExtendedPanId;    ///< Extended PAN ID
-    otMeshLocalPrefix              mMeshLocalPrefix;  ///< Mesh Local Prefix
-    uint32_t                       mDelay;            ///< Delay Timer
-    otPanId                        mPanId;            ///< PAN ID
-    uint16_t                       mChannel;          ///< Channel
-    otPskc                         mPskc;             ///< PSKc
-    otSecurityPolicy               mSecurityPolicy;   ///< Security Policy
-    otChannelMask                  mChannelMask;      ///< Channel Mask
-    otOperationalDatasetComponents mComponents;       ///< Specifies which components are set in the Dataset.
+    uint64_t          mActiveTimestamp;  ///< Active Timestamp (value: (seconds << 16) | (ticks << 1) | authoritive))
+    uint64_t          mPendingTimestamp; ///< Pending Timestamp (value: (seconds << 16) | (ticks << 1) | authoritive))
```

Comment:
In general, I think adding more accuracy (ticks) to the timestamp sounds reasonable.

But I am concerned that this change as it is  can impact any existing next layer code that was using these APIs and definitions. We are changing the meaning of `mActiveTimestamp` and `mPendingTimestamp` (it breaks backward compatibility).

Some questions/ideas:
- First question if we do we need the ticks level accuracy or not. Can we continue with the existing seconds definition?
- If we want to change, it may be good to make the new definition/API very different to help make it to noticable for any current users of the API (avoid compatibility issues):
   - We can keep `mActiveTimestamp` as before (seconds value) and add additional member variables for ticks, etc (though it may still break any current user of the APIs)    
  - The other way may be to rename the the fields  to make it different from its current name `mActiveTimestamp` (make it noticeable for any next layer code that is using these APIs/definitions that it is changed so they can adapt accordingly).
  - Another way may be to keep the member variable name same but define it as a new type (say we define a `struct` representing a timestamp as separate seconds, ticks and auth flags). I think this way we again make it noticeable for any user of current API that things have changed.

Thoughts?

Reply:
> First question if we do we need the ticks level accuracy or not. Can we continue with the existing seconds definition?

The problem with current definition is that upper layer cannot get the complete Timestamp values. For example, Thread spec requires the MeshCoP mDNS Service to advertise the full Timestamp in the TXT data (keyed by `at`). If the user match the the Active Timestamp in mDNS service against the Active Dataset cache on mobile to test if the Active Dataset is up-to-date, it will always get a false result. It also force a Border Router implementation to violates the Thread spec (or do we want to change the spec?).

> Another way may be to keep the member variable name same but define it as a new type (say we define a struct representing a timestamp as separate seconds, ticks and auth flags). I think this way we again make it noticeable for any user of current API that things have changed.

This sounds good to me. @librasungirl @simonlingoogle

Reply:
I also prefer the third option (defining new type) over the others.

Reply:
@abtink Updated to to use a new `otTimestamp` struct, PTAL.